File: src/main/java/com/braydenoneal/blang/parser/expression/value/ItemStackValue.java
package com.braydenoneal.blang.parser.expression.value;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.item.ItemStack;

public class ItemStackValue extends Value<ItemStack> {
    public static final MapCodec<ItemStackValue> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
            ItemStack.CODEC.fieldOf("value").forGetter(ItemStackValue::value)
    ).apply(instance, ItemStackValue::new));

    public ItemStackValue(ItemStack value) {
        super(value);
    }

    @Override
    public ValueType<?> getType() {
        return ValueTypes.ITEM_STACK;
    }

    @Override
    public boolean equals(Object obj) {
        try {
            if (obj instanceof ItemStackValue itemValue) {
                return value().equals(itemValue.value());
            }
        } catch (Error e) {
            System.out.println("Cannot equate item stack values outside of the game");
        }

        return false;
    }
}

File: src/main/kotlin/com/braydenoneal/blang/parser/expression/value/ItemStackValue.kt
package com.braydenoneal.blang.parser.expression.value

import com.braydenoneal.blang.parser.RunException
import com.mojang.serialization.MapCodec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.item.ItemStack

class ItemStackValue(value: ItemStack) : Value<ItemStack>(value) {
    override val valueType: ValueType<*> get() = ValueTypes.ITEM_STACK

    override fun equals(other: Any?): Boolean {
        try {
            if (other is ItemStackValue) {
                return value == other.value
            }
        } catch (_: Error) {
            throw RunException("Cannot equate item stack values outside of the game")
        }

        return false
    }

    companion object {
        val CODEC: MapCodec<ItemStackValue> = RecordCodecBuilder.mapCodec { instance ->
            instance.group(
                ItemStack.CODEC.fieldOf("value").forGetter(ItemStackValue::value)
            ).apply(instance, ::ItemStackValue)
        }
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + valueType.hashCode()
        return result
    }
}
